File: src/expr.rs
use crate::token::Token;

pub struct BinaryExpr {
    pub left: Box<Expr>,
    pub operator: Token,
    pub right: Box<Expr>,
}

pub enum LiteralValue {
    Number(f64),
    Str(String),
    Bool(bool),
    Nil,
}

pub enum Expr {
    Assign {
        name: Token,
        value: Box<Expr>,
    },
    Binary {
        left: Box<Expr>,
        operator: Token,
        right: Box<Expr>,
    },
    Call {
        callee: Box<Expr>,
        paren: Token,
        arguments: Vec<Token>,
    },
    Get {
        object: Box<Expr>,
        name: Token,
    },
    Grouping {
        expression: Box<Expr>,
    },
    Literal {
        value: LiteralValue,
    },
    Logical {
        left: Box<Expr>,
        operator: Token,
        right: Box<Expr>,
    },
    Set {
        object: Box<Expr>,
        name: Token,
        value: Box<Expr>,
    },
    Super {
        keyword: Token,
        method: Token,
    },
    This {
        keyword: Token,
    },
    Unary {
        operator: Token,
        right: Box<Expr>,
    },
    Variable {
        name: Token,
    },
}

pub trait Visitor<R> {
    fn visit_assign_expr(&mut self, expr: &Expr) -> R;
    fn visit_binary_expr(&mut self, expr: &Expr) -> R;
    fn visit_call_expr(&mut self, expr: &Expr) -> R;
    fn visit_get_expr(&mut self, expr: &Expr) -> R;
    fn visit_grouping_expr(&mut self, expr: &Expr) -> R;
    fn visit_literal_expr(&mut self, expr: &Expr) -> R;
    fn visit_logical_expr(&mut self, expr: &Expr) -> R;
    fn visit_set_expr(&mut self, expr: &Expr) -> R;
    fn visit_super_expr(&mut self, expr: &Expr) -> R;
    fn visit_this_expr(&mut self, expr: &Expr) -> R;
    fn visit_unary_expr(&mut self, expr: &Expr) -> R;
    fn visit_variable_expr(&mut self, expr: &Expr) -> R;
}

impl Expr {
    pub fn accept<R>(&self, visitor: &mut dyn Visitor<R>) -> R {
        match self {
            Expr::Assign { name, value } => visitor.visit_assign_expr(&self),
            Expr::Binary { left, operator, right } => visitor.visit_binary_expr(&self),
            Expr::Call { callee, paren, arguments } => visitor.visit_call_expr(&self),
            Expr::Get { object, name } => visitor.visit_get_expr(&self),
            Expr::Grouping { expression } => visitor.visit_grouping_expr(&self),
            Expr::Literal { value } => visitor.visit_literal_expr(&self),
            Expr::Logical { left, operator, right } => visitor.visit_logical_expr(&self),
            Expr::Set { object, name, value } => visitor.visit_set_expr(&self),
            Expr::Super { keyword, method } => visitor.visit_super_expr(&self),
            Expr::This { keyword } => visitor.visit_this_expr(&self),
            Expr::Unary { operator, right } => visitor.visit_unary_expr(&self),
            Expr::Variable { name } => visitor.visit_variable_expr(&self),
        }
    }
}

File: src/main/java/com/craftinginterpreters/lox/Expr.java
package com.craftinginterpreters.lox;

import java.util.List;

abstract class Expr {

	interface Visitor<R> {
		R visitBinaryExpr(Binary expr);
		R visitGroupingExpr(Grouping expr);
		R visitLiteralExpr(Literal expr);
		R visitUnaryExpr(Unary expr);
	}

	abstract <R> R accept(Visitor<R> visitor);

	static class Binary extends Expr {

		public final Expr left;
		public final Token operator;
		public final Expr right;

		public Binary(Expr left, Token operator, Expr right) {
			this.left = left;
			this.operator = operator;
			this.right = right;
		}

		@Override
		 <R> R accept(Visitor<R> visitor) {
			return visitor.visitBinaryExpr(this);
		}
	}

	static class Grouping extends Expr {

		public final Expr expression;

		public Grouping(Expr expression) {
			this.expression = expression;
		}

		@Override
		 <R> R accept(Visitor<R> visitor) {
			return visitor.visitGroupingExpr(this);
		}
	}

	static class Literal extends Expr {

		public final Object value;

		public Literal(Object value) {
			this.value = value;
		}

		@Override
		 <R> R accept(Visitor<R> visitor) {
			return visitor.visitLiteralExpr(this);
		}
	}

	static class Unary extends Expr {

		public final Token operator;
		public final Expr right;

		public Unary(Token operator, Expr right) {
			this.operator = operator;
			this.right = right;
		}

		@Override
		 <R> R accept(Visitor<R> visitor) {
			return visitor.visitUnaryExpr(this);
		}
	}
}
